src/hotspot/share/runtime/thread.cpp

```diff
@@ -1364,6 +1364,7 @@ static bool is_daemon(oop threadObj) {
 // cleanup_failed_attach_current_thread as well.
 void JavaThread::exit(bool destroy_vm, ExitType exit_type) {
   assert(this == JavaThread::current(), "thread consistency check");
+  assert(!is_exiting(), "should not be exiting or terminated already");
```

Comment:
Hmmm... The old `destroy_vm == true` branch assert was:

`assert(!is_terminated() && !is_exiting(), "must not be exiting");`

and the new assert mesg has "or terminated", but the assert is only
checking `!is_exiting()` and doesn't not include `!is_terminated()`.
So the old assert checked both conditions and only mumbled about
"exiting" and the new assert checks one condition and mumbles
about both conditions.

Reply:
is_exiting() also checks is_terminated() internally, so it returns true for the _thread_exiting, _thread_terminated and _vm_exited cases (false only for the _not_terminated case). Since the extra !is_terminated() check was redundant I removed it. Also since the assert will trigger for both the exiting and terminated cases I added both to the assert msg.